File: apps/go-core/internal/service/business.service.go
package service

import (
	"context"
	"fmt"
	"log/slog"

	"github.com/bimal009/Zovly/internal/models"
	repository "github.com/bimal009/Zovly/internal/repo"
	"github.com/jmoiron/sqlx"
)

type BusinessService interface {
	Create(ctx context.Context, input models.Business, userId string) (*models.Business, error)
	GetByUserId(ctx context.Context, userId string) (*models.BusinessWithMembers, error)
}

type businessService struct {
	db                 *sqlx.DB
	businessRepo       repository.BusinessRepo
	businessMemberRepo repository.BusinessMemberRepo
	userRepo           repository.UserRepo
	appRepo            repository.AppRepo
	log                *slog.Logger
}

func NewBusinessService(
	db *sqlx.DB,
	businessRepo repository.BusinessRepo,
	businessMemberRepo repository.BusinessMemberRepo,
	userRepo repository.UserRepo,
	log *slog.Logger,
	appRepo repository.AppRepo,
) BusinessService {
	return &businessService{
		db:                 db,
		businessRepo:       businessRepo,
		businessMemberRepo: businessMemberRepo,
		userRepo:           userRepo,
		log:                log,
		appRepo:            appRepo,
	}
}

func ptr[T any](v T) *T { return &v }

func (s *businessService) Create(ctx context.Context, input models.Business, userId string) (*models.Business, error) {
	tx, err := s.db.BeginTxx(ctx, nil)
	if err != nil {
		return nil, fmt.Errorf("begin tx: %w", err)
	}
	defer tx.Rollback()

	created, err := s.businessRepo.CreateTx(ctx, tx, input)
	if err != nil {
		return nil, fmt.Errorf("insert: %w", err)
	}

	member := models.BusinessMember{
		BusinessID: created.ID,
		UserID:     userId,
		Role:       models.MemberRoleOwner,

		CanManageContent:   true,
		CanViewAnalytics:   true,
		CanManageAds:       true,
		CanReadDMs:         true,
		CanReplyDMs:        true,
		CanReadComments:    true,
		CanReplyComments:   true,
		CanViewLeads:       true,
		CanManageLeads:     true,
		CanViewBookings:    true,
		CanManageBookings:  true,
		CanViewInventory:   true,
		CanManageInventory: true,
		CanViewOrders:      true,
		CanManageSettings:  true,
		CanManageMembers:   true,
		CanManageBilling:   true,
	}

	_, err = s.businessMemberRepo.CreateTx(ctx, tx, member)
	if err != nil {
		return nil, fmt.Errorf("create business member: %w", err)
	}

	if err = s.appRepo.Create(ctx, tx, created.ID); err != nil {
		return nil, fmt.Errorf("create app connections: %w", err)
	}

	_, err = s.userRepo.UpdateTx(ctx, tx, userId, models.UserUpdate{
		Role:      ptr(models.RoleVendor),
		Onboarded: ptr(true),
	})
	if err != nil {
		return nil, fmt.Errorf("update user: %w", err)
	}

	if err = tx.Commit(); err != nil {
		return nil, fmt.Errorf("commit: %w", err)
	}

	s.log.InfoContext(ctx,
		"business created",
		"business_id", created.ID,
		"owner_user_id", userId,
	)

	return created, nil
}

func (s *businessService) GetByUserId(ctx context.Context, userId string) (*models.BusinessWithMembers, error) {
	result, err := s.businessRepo.GetByUserId(ctx, userId)
	if err != nil {
		return nil, fmt.Errorf("get business: %w", err)
	}
	return result, nil
}

File: apps/api-v1/src/modules/business/business.service.ts
import { eq } from "drizzle-orm";
import { CreateBusinessInput } from "@repo/types";

import { db } from "../../config/db/db";
import { business } from "../../config/db/schema/business";
import { auth, User } from "../../lib/auth";
import { businessMembers } from "../../config/db/schema/members";
import {
  ConflictError,
  InternalServerError,
  NotFoundError,
  UnauthorizedError,
} from "../../lib/errors";
import { getByUserId } from "../business members/members.service";

export const create = async (
  session: User,
  input: CreateBusinessInput,
  headers: Headers,
) => {
  if (!session) {
    throw new UnauthorizedError("User not found");
  }

  const existingMember = await getByUserId(session.id);

  if (existingMember) {
    throw new ConflictError("User already has a business.");
  }
  console.log(existingMember);

  const newBusiness = await db.transaction(async (tx) => {
    const [newBusiness] = await tx.insert(business).values(input).returning();

    await tx.insert(businessMembers).values({
      businessId: newBusiness.id,
      userId: session.id,
      role: "owner",
      canManageAds: true,
      canManageBilling: true,
      canManageBookings: true,
      canManageContent: true,
      canManageInventory: true,
      canManageLeads: true,
      canManageMembers: true,
      canManageSettings: true,
      canReadComments: true,
      canReadDms: true,
      canReplyComments: true,
      canReplyDms: true,
      canViewAnalytics: true,
      canViewBookings: true,
      canViewInventory: true,
      canViewLeads: true,
      canViewOrders: true,
      canEditBusiness:true,
      
    });

    try {
      await auth.api.updateUser({
        body: {
          isOnboarded: true,
          role: "vendor",
        },
        headers,
      });
    } catch (err) {
      throw new InternalServerError(
        "Failed to update user session. Please try again. Error: " + err,
      );
    }

    return newBusiness;
  });

  return newBusiness;
};

export const getById = async (id: string) => {
  const record = await db.query.business.findFirst({
    where: eq(business.id, id),
  });

  if (!record) {
    throw new NotFoundError("Business not found");
  }

  return record;
};
